File: ios/Classes/ChatSdkFlutterPlugin.swift
import BrrmChatFramework
import Flutter
import UIKit
public class ChatSdkFlutterPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "chat_sdk_flutter", binaryMessenger: registrar.messenger())
        let instance = ChatSdkFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
        case Methods.INIT_CHAT.rawValue:
            guard let args = call.arguments as? [String: Any] else { return }
            guard let token = args["appToken"] as? String else { return }
            BrrmChat.shared.setToken(applicationToken: token)
            result(true)
        case Methods.REGISTER.rawValue:
            guard let args = call.arguments as? [String: Any] else {
                result(false)
                return
            }
            guard let userMap = args["user"] as? [String: Any] else {
                result(false)
                return
            }
            guard let groupMap = args["group"] as? [String: Any] else {
                result(false)
                return
            }
            guard let brrmUser = BrrmUser(json: userMap), let brrmGroup = BrrmGroup(json: groupMap) else {
                result(false)
                return
            }
            
            let fcmToken = args["FCMToken"]

            BrrmChat.shared.register(user: brrmUser, group: brrmGroup,fcmToken:fcmToken)
            result(true)
        case Methods.OPEN_CHAT.rawValue:
            BrrmChat.shared.openChatList()
            result(true)
        case Methods.SET_FCM_TOKEN.rawValue:
            guard let args = call.arguments as? [String: Any], let fcmToken = args["FCMToken"] as? String else {
                result(false)
                return
            }
            BrrmChat.shared.setFMCToken(fmcToken: fcmToken)
            result(true)
        case Methods.HANDLE_BRRM_CHAT_MESSAGE.rawValue:
            guard let args = call.arguments as? [String: Any] else {
                result(false)
                return
            }
            if BrrmChat.shared.isBrrmChatNotification(userInfo: args) {
                BrrmChat.shared.notification(userInfo: args)
                result(true)
            }
            result(false)
        case Methods.IS_BRRM_CHAT_MESSAGE.rawValue:
            guard let args = call.arguments as? [String: Any] else {
                result(false)
                return
            }
            result(BrrmChat.shared.isBrrmChatNotification(userInfo: args))

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

enum Methods: String {
    case INIT_CHAT = "initChat"
    case OPEN_CHAT = "openChat"
    case SET_FCM_TOKEN = "setFCMToken"
    case HANDLE_BRRM_CHAT_MESSAGE = "handleBrrmChatMessage"
    case IS_BRRM_CHAT_MESSAGE = "isBrrmChatMessage"

    case REGISTER = "register"
}

File: android/src/main/kotlin/eu/brrm/chat_sdk_flutter/ChatSdkFlutterPlugin.kt
package eu.brrm.chat_sdk_flutter

import android.content.Context
import eu.brrm.chatui.BrrmChat
import eu.brrm.chatui.internal.ChatEnvironment
import eu.brrm.chatui.internal.data.BrrmGroup
import eu.brrm.chatui.internal.data.BrrmUser
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONObject

/** ChatSdkFlutterPlugin */
class ChatSdkFlutterPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    private var context: Context? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "chat_sdk_flutter")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            Methods.GET_PLATFORM_VERSION -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }

            Methods.INIT -> {
                init(call, result)
            }

            Methods.OPEN_CHAT -> {
                openChat(call, result)
            }

            Methods.SET_FCM_TOKEN -> {
                setFcmToken(call, result)
            }

            Methods.REGISTER -> {
                register(call, result)
            }

            Methods.IS_BRRM_CHAT_MESSAGE -> {
                isBrrmChatMessage(call, result)
            }

            Methods.HANDLE_BRRM_CHAT_MESSAGE -> {
                handleBrrmChatMessage(call, result)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun handleBrrmChatMessage(call: MethodCall, result: Result) {
        val data = call.arguments as? Map<*, *>

        data?.let {
            BrrmChat.instance.handleBrrmChatMessage(it)
        }
        result.success(true)
    }

    private fun isBrrmChatMessage(call: MethodCall, result: Result) {
        val data = call.arguments as? Map<*, *>
        val isBrrmChatMessage = data?.let { BrrmChat.instance.isBrrmChatMessage(it) } ?: false
        result.success(isBrrmChatMessage)
    }

    private fun register(call: MethodCall, result: Result) {
        val user = call.argument<Map<*, *>>("user")?.let { BrrmUser.fromJSON(JSONObject(it)) }
        val group = call.argument<Map<*, *>>("group")?.let { BrrmGroup.fromJSON(JSONObject(it)) }
        val fcmToken = call.argument<String>("FCMToken")
        if (user != null && group != null) {
            BrrmChat.instance.register(user, group, fcmToken) {
                result.success(it)
            }
        }
    }

    private fun setFcmToken(call: MethodCall, result: Result) {
        call.argument<String>("FCMToken")?.let { token ->
            BrrmChat.instance.subscribeDevice(token) {
                result.success(it)
            }
        }
    }

    private fun init(call: MethodCall, result: Result) {
        context?.let { ctx ->
            val args = call.arguments as Map<*, *>
            val appToken = args["appToken"] as? String
                ?: throw IllegalArgumentException("AppToken must be provided")
            val chatEnv = (args["chatEnv"] as? Int).let {
                ChatEnvironment.entries[it ?: 0]
            }
            BrrmChat.init(ctx, appToken, chatEnv)
        }
        result.success(true)
    }

    private fun openChat(call: MethodCall, result: Result) {
        context?.let {
            BrrmChat.instance.openChatList(it)
        }
        result.success(true)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = null
        channel.setMethodCallHandler(null)
    }
}
